### ISSUE TITLE
Serialization sequence normalization step 3 needs clarification

### ISSUE THREAD

**Issue by Arithmeticus** [qt4cg/qtspecs#464]
The specifications currently read:

> If the item-separator serialization parameter is absent, then for each subsequence of adjacent strings in S2, copy a single string to the new sequence equal to the values of the strings in the subsequence concatenated in order, each separated by a single space. Copy all other items to the new sequence. Otherwise, copy each item in S2 to the new sequence, inserting between each pair of items a string whose value is equal to the value of the item-separator parameter. The new sequence is S3.

As written ("If...then.... Otherwise...."), this implies that the process whereby a block of adjacent strings are joined into a single string is performed only when the item separator parameter is absent. I.e., if the item-separator parameter is not absent, it will not be used to string-join adjacent groups of strings. Perhaps that is as intended, but I want to make sure.

Also, as written, this implies that when the parameter is absent, the sequence begins by finding all subsequences of adjacent strings, performing concatenation. Then all non-strings are added to the sequence. That seems wrong, because it appears to advise the processor to rearrange the sequence of input items.

I propose a revision along these lines:
> Copy each item in S2 to a new sequence. If a given pair of adjacent items are both strings, then separate them with a string whose value is equal to the value of the item-separator parameter or is a single space if the item-separator parameter is absent. If a given pair of adjacent items are not both strings, insert between each pair of items a string whose value is equal to the value of the item-separator parameter. Once this is finished, take each adjacent group of strings and concatenate them into a single string. The new sequence is S3.

My revision is based upon what I imagine happens, but implementers will know better than I.

I am working on some editorial touch-ups of the Serialization specifications, and can incorporate comments/suggestions made in this thread in that larger enterprise.

**Comment by michaelhkay** [comment]
I agree this text can easily be misread, and I don't think your reading is quite correct.

It's really easiest to explain this by example, and the spec could probably do with more examples.

Consider the query

<a/>,<b/>,3 to 5,<c/>,6 to 8

With an item separator = "|" it serializes as

<a/>|<b/>|3|4|5|<c/>|6|7|8

Here the separator is inserted between all items, including nodes as well as atomic values. The sequence S3 contains 17 items, 9 from the original sequence, and 8 separators.

without an item-separator, this serializes (with the XML method) as

<a/><b/>3 4 5<c/>6 7 8

that is, a single space is inserted between adjacent atomic values (which by now have been converted to strings).

The sequence S3 in this case consists of 5 items: the elements a and b, the string "3 4 5", the element c, and the string "6 7 8"

Now, does it actually matter that S3 consists of 5 items, rather than 13 (9 from the original sequence, plus 4 separators). I suspect it doesn't, because if S3 had 13 items, rule 6 would concatenate the strings into text nodes anyway, so sequence S6 would still have 5 items (3 elements and 2 text nodes). But the text here is fragile and I'm inclined towards minimal change.

Note that rule 6 has the same problem as rule 3 - it could be read as suggesting that the grouped items are copied to the result sequence first, followed by the ungrouped items.

How to improve the explanation in the spec? Perhaps for rule 3:

* If the item-separator serialization parameter is present, then copy each item in S2 to the new sequence, inserting between each pair of items a string whose value is equal to the value of the item-separator parameter.

* Otherwise (when the item-separator parameter is absent): first group the items in S2 into a sequence of groups, where each group consists either of a single item that is not an xs:string, or of a maximal sequence of items of type xs:string. For a group of the first kind (a single non-string item), copy it to the new sequence unchanged. For a group of the second kind (a sequence of xs:string values), construct a single string equal to the values of the strings in the group concatenated in order, adjacent strings being separated by a single space; then copy this constructed string to the new sequence.

And similarly for rule 6.

I find it a little surprising that rule 7 says "which are all nodes". What happened to any maps or other functions (other than arrays) that might have been in the input sequence? The example at the end of section 2 says "This process results in a serialization error [err:SENR0001] if $seq contains functions, attribute nodes or namespace nodes."; but SENR0001 only mentions namespace and attribute nodes.

Have fun sorting this out...

Michael Kay
Saxonica

**Comment by Arithmeticus** [comment]
After posting PR #555 I continued reading, and I have to wonder about the lengthy note that follows ([here](https://qt4cg.org/specifications/xslt-xquery-serialization-40/Overview-diff.html#serdm), scroll down). It does not specify whether it describes what happens in all seven steps, or if it is focused only on step 3. I'd like to assume S3, because it begins with a prominent discussion of `item-separator` and shows no interest in the `array:flatten()` step (S1). But because the sample instructions encase the results in a document node, the whole process through S7 might be in mind, attempting to compress all seven steps into a single XSLT/XQuery expression. 

Regardless, the XSLT and XQuery appear wrong. S7's document node can contain (to that point) intact maps, which would be lost with `<xsl:value-of select="."/>` (line 12 in the XSLT instruction). If these XSLT/XQuery instructions are focused on describing the creation of S3, the choose-when-otherwise should put the `xs:string` test in the `when` and the rest in the `otherwise`. If the instructions focus on S7, then maps/arrays need to be accommodated.

I'm not in the best position to revise these code snippets. OTOH, if comments to this thread help me better understand what is being written, hopefully I can use that to revise the note's prose.

PR #555 does not address the question of whether a function should trigger `err:SENR0001`, which is stated twice, once in the step descriptions (of S6) and again in the lengthy note (incorrectly of S0, or S2, depending upon the scope  of the note). As noted by @michaelhkay the error's definition is restricted to namespace nodes and attribute nodes only.

**Comment by Arithmeticus** [comment]
Closed with #555
